### ISSUE TITLE
3D map freezes when camera is adjusted against map bounds (in dev environment only (?))

### ISSUE THREAD

**Issue by larsmaxfield** [maplibre/maplibre-gl-js#4537]
**maplibre-gl-js version**: 4.5.1

**browser**: Edge 126.0.2592.102 (64-bit), Firefox 128.0 (64-bit)

**Development environment** on Windows 10 WSL

### Steps to Trigger Behavior

1. In a development environment, open a 3D terrain map (`npm run start`; `localhost:XXXX/test/examples/3d-terrain.html`).
2. Enable 3D and pitch the map.
3. Pan to the map's latitude bounds.
4. Adjust the camera (pan, zoom, pitch, rotate) against the bounds.
5. Map freezes due to uncaught error in `CanonicalTileID` with zxy out of bounds.

![map-freeze](https://github.com/user-attachments/assets/ae70432e-e5e8-46c4-af55-6d0479361895)

![uncaught-error](https://github.com/user-attachments/assets/44ee4623-a86a-426d-acb6-55c4dd35f385)

### Link to Demonstration

I can't seem to trigger this in the [online docs](https://maplibre.org/maplibre-gl-js/docs/examples/3d-terrain.html#2.38/74.02/-138.93/34.8/82) — only in a local development environment.

### Expected Behavior

The map should not freeze when adjusting the camera against the map bounds.

### Actual Behavior

The map freezes.

**Comment by larsmaxfield** [comment]
The issue is occurring because `getElevationForLngLatZoom` is making requests to `OverscaledTileID` with LngLats outside the map bounds:

https://github.com/maplibre/maplibre-gl-js/blob/22075197c611f887d3b7ee03fe136a0f142fc4b1/src/render/terrain.ts#L186-L189

This can be fixed by first checking if the LngLat and zoom are in bounds before the request to the `OverscaledTileID`. If it's out of bounds, the function returns an elevation of 0:

https://github.com/maplibre/maplibre-gl-js/blob/f850046882cd5e7a8ff3e5eba54e6fc80d51fa1a/src/render/terrain.ts#L186-L191

**Comment by HarelM** [comment]
Feel free to submit a PR.
If there's a check already, I would prefer to have it somewhere centralized (utils.ts maybe) so that changing it once will work for all the relevant places.
Especially if we ever decide to change the max zoom or min zoom.

**Comment by larsmaxfield** [comment]
Good point. 

I centralized the new constants `MAX_TILE_ZOOM = 25` and `MIN_TILE_ZOOM = 0` in `util.ts`. I found a couple instances where those were hardcoded and replaced them accordingly.

However, I kept the bounds check functions in `tile_id.ts`. The check `isInBoundsForZoomLngLat` (which is used to fix this issue) relies on `LngLat`. If I put that in `util.ts`, it caused a circular dependency.
